scripts/data/gen-nodetime/src/relayer/lib/relayer.ts

```diff
@@ -0,0 +1,433 @@
+		try {
+			const tmClient = await Tendermint34Client.connect(rpcAddr);
+			let status = await tmClient.status();
+			const chain = {
+				chainId: status.nodeInfo.network,
+				rpcAddr,
+				addrPrefix,
+				gasPrice,
+			};
+			if (!this.config.chains) {
+				this.config.chains = [chain];
+				return { id: chain.chainId };
+			}
+			const endpointExistsWithDifferentChainID =
+				this.config.chains &&
+				this.config.chains.find(
+					(x) => x.chainId != chain.chainId && x.rpcAddr == chain.rpcAddr
+				);
+			if (endpointExistsWithDifferentChainID)
+				throw new Error(Errors.endpointExistsWithDifferentChainID);
+
+			const chainExistsWithSameEndpoint =
+				this.config.chains &&
+				this.config.chains.find(
+					(x) => x.chainId == chain.chainId && x.rpcAddr == chain.rpcAddr
+				);
+			if (chainExistsWithSameEndpoint) return { id: chain.chainId };
+			this.config.chains.push(chain);
+			return { id: chain.chainId };
+		} catch (e) {
+			throw new Error(Errors.chainSetupFailed + e);
+		}
+	}
+	public createPath([srcID, dstID, options]: [
+		string,
+		string,
+		ConnectOptions
+	]): Path {
+		try {
+			let path = {
+				id: srcID + "-" + dstID,
+				isLinked: false,
+				src: {
+					chainID: srcID,
+					portID: options.sourcePort,
+				},
+				dst: {
+					chainID: dstID,
+					portID: options.targetPort,
+				},
+			};
+			if (!this.config.paths) {
+				this.config.paths = [];
+			}
+			this.config.paths.push({ path, options });
+			return path;
+		} catch (e) {
+			throw new Error(Errors.pathSetupFailed + e);
+		}
+	}
+	public getPath([id]: [string]): Path {
+		if (this.config.paths) {
+			let path = this.config.paths.find((x) => x.path.id == id);
+			if (path) return path.path;
+		}
+		throw new Error(Errors.pathNotExists);
+	}
+	public listPaths(): Path[] {
+		if (this.config.paths) {
+			let paths = this.config.paths.map((x) => x.path);
+			return paths;
+		}
+		throw new Error(Errors.pathsNotDefined);
+	}
+	public async getDefaultAccount([chainID]: string[]): Promise<Account> {
+		const chain = this.chainById(chainID);
+		if (chain) {
+			let client = await this.getIBCClient(chain);
+			return {
+				address: client.senderAddress,
+			};
+		}
+		throw new Error(Errors.chainNotFound + chainID);
+	}
+	public async getDefaultAccountBalance([chainID]: string[]): Promise<Coin[]> {
+		const chain = this.chainById(chainID);
+		if (chain) {
+			let client = await this.getIBCClient(chain);
+			return await client.query.bank.allBalances(client.senderAddress);
+		}
+		throw new Error(Errors.chainNotFound + chainID);
+	}
+	public async link(paths: string[]): Promise<LinkResponse> {
+		if (this.config.paths) {
+			let response: LinkResponse = {
+				linkedPaths: [],
+				alreadyLinkedPaths: [],
+			};
+			for (let pathName of paths) {
+				const path = this.pathById(pathName);
+				if (path?.path.isLinked) {
+					response.alreadyLinkedPaths.push(pathName);
+					continue;
+				}
+				try {
+					await this.createLink(path);
```

Comment:
we should run this concurrently for all paths otherwise link() will be very slow.

Reply:
We attempted this but it causes issues. Let's stick to the previous behaviour